lib/galaxy/webapps/galaxy/api/workflows.py

```diff
@@ -603,6 +609,168 @@ def build_module(self, trans, payload={}):
         }
 
+    @expose_api
+    def get_tool_predictions(self, trans, payload, **kwd):
+        """
+        POST /api/workflows/get_tool_predictions
+        Fetch predicted tools for a workflow
+        """
+        if 'tool_sequence' not in payload or trans.app.config.tool_recommendation_model_path is None:
+            return
+        tool_sequence = payload.get('tool_sequence', "")
+        self.enable_admin_tool_recommendations = trans.app.config.enable_admin_tool_recommendations
+        # collect tool recommendations if set by admin
+        if not self.admin_tool_recommendations_path:
+            if self.enable_admin_tool_recommendations is True or self.enable_admin_tool_recommendations == "true":
+                self.admin_tool_recommendations_path = os.path.join(os.getcwd(), trans.app.config.admin_tool_recommendations_path)
+                with open(self.admin_tool_recommendations_path) as admin_recommendations:
+                    self.admin_recommendations_list = json.loads(admin_recommendations.read())
+        # get model
+        if not self.tool_recommendation_model_path:
+            self.tool_recommendation_model_path = os.path.join(os.getcwd(), trans.app.config.tool_recommendation_model_path)
+            self.all_tools = dict()
+            # collect ids and names of all the installed tools
+            for tool_id, tool in trans.app.toolbox.tools():
+                t_id_renamed = tool_id
+                if t_id_renamed.find("/") > -1:
+                    t_id_renamed = t_id_renamed.split("/")[-2]
+                self.all_tools[t_id_renamed] = (tool_id, tool.name)
+            self.trained_model = h5py.File(self.tool_recommendation_model_path, 'r')
+            self.model_config = json.loads(self.trained_model.get('model_config').value)
+            self.loaded_model = model_from_json(self.model_config)
+            self.model_data_dictionary = json.loads(self.trained_model.get('data_dictionary').value)
+            self.compatible_tools = json.loads(self.trained_model.get('compatible_tools').value)
+            self.tool_weights = json.loads(self.trained_model.get('class_weights').value)
+            self.tool_weights_sorted = dict()
+            tool_pos_sorted = [int(key) for key in self.tool_weights.keys()]
+            for k in tool_pos_sorted:
+                self.tool_weights_sorted[k] = self.tool_weights[str(k)]
+            self.reverse_dictionary = dict((v, k) for k, v in self.model_data_dictionary.items())
+        model_weights = list()
+        weight_ctr = 0
+        while True:
+            try:
+                d_key = "weight_" + str(weight_ctr)
+                weights = self.trained_model.get(d_key).value
+                model_weights.append(weights)
+                weight_ctr += 1
+            except Exception as exception:
+                break
+        # set the model weights
+        self.loaded_model.set_weights(model_weights)
+        tool_sequence = tool_sequence.split(",")
```

Comment:
This would give you a list with an empty string if `tool_sequence` is '' - which, I guess, is not what you want here. You could use `[]` as the default in line 620. Then instead of lines 620, 661 and 662, you could have this:
```python
tool_sequence = payload.get('tool_sequence', [])
if tool_sequence:
    tool_sequence = tool_sequence.split(',')[::-1]
```
This will give you a reversed list or an empty list.
These are all just minor suggestions.